src/main/java/com/yelp/nrtsearch/server/luceneserver/field/IndexableFieldDef.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2020 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.field;
+
+import com.yelp.nrtsearch.server.grpc.Field;
+import com.yelp.nrtsearch.server.luceneserver.ServerCodec;
+import com.yelp.nrtsearch.server.luceneserver.doc.LoadedDocValues;
+import java.io.IOException;
+import java.util.List;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.similarities.BM25Similarity;
+import org.apache.lucene.search.similarities.ClassicSimilarity;
+import org.apache.lucene.search.similarities.Similarity;
+
+/** Base class for all field definition types that can be written into the index. */
+public abstract class IndexableFieldDef extends FieldDef {
```

Comment:
could this be an interface with `default` methods instead?

Reply:
You could change them to interfaces, but the default functions wouldn't be helpful in most cases. Since the interface can't store state, every type would need to redefine the standard functions (getName/hasDocValues/isMultiValue/etc) and do its own grpc Field definition processing. It felt like it was going to be a lot of code duplication, so I was trying to use the abstract class to structure the field building through its protected methods. I am open to discussion on this.